File: enterprise-starter-kit/src/main/java/com/enterprise/starter/kit/modules/auth/service/ProfileService.java
package com.enterprise.starter.kit.modules.auth.service;

import com.enterprise.starter.kit.modules.auth.dto.ChangePasswordRequest;
import com.enterprise.starter.kit.modules.auth.dto.ProfileResponse;
import com.enterprise.starter.kit.modules.auth.dto.ProfileUpdateRequest;
import com.enterprise.starter.kit.modules.auth.entity.User;
import com.enterprise.starter.kit.modules.auth.repository.UserRepository;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service for managing user profile operations.
 * Handles profile retrieval, updates, and password changes.
 */
@Service
public class ProfileService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public ProfileService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Gets the currently authenticated user from the security context.
     *
     * @return the authenticated User entity
     * @throws IllegalStateException if no user is authenticated
     */
    private User getCurrentUser() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal instanceof User user) {
            // Refresh from database to get latest data
            return userRepository.findById(user.getId())
                    .orElseThrow(() -> new IllegalStateException("Current user not found in database"));
        }
        throw new IllegalStateException("No authenticated user found");
    }

    /**
     * Retrieves the current user's profile information.
     *
     * @return ProfileResponse containing user details
     */
    @Transactional(readOnly = true)
    public ProfileResponse getProfile() {
        User user = getCurrentUser();
        return new ProfileResponse(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getFullName(),
                user.getPhoneNumber(),
                user.getAvatarUrl(),
                user.getPreferredLanguage()
        );
    }

    /**
     * Updates the current user's profile information.
     *
     * @param request the profile update data
     * @return updated ProfileResponse
     */
    @Transactional
    public ProfileResponse updateProfile(ProfileUpdateRequest request) {
        User user = getCurrentUser();

        // Update fields if provided
        if (request.fullName() != null) {
            user.setFullName(request.fullName());
        }
        if (request.phoneNumber() != null) {
            user.setPhoneNumber(request.phoneNumber());
        }
        if (request.preferredLanguage() != null) {
            user.setPreferredLanguage(request.preferredLanguage());
        }

        userRepository.save(user);

        return new ProfileResponse(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getFullName(),
                user.getPhoneNumber(),
                user.getAvatarUrl(),
                user.getPreferredLanguage()
        );
    }

    /**
     * Changes the current user's password.
     * Validates the old password before updating.
     *
     * @param request containing old and new passwords
     * @throws IllegalArgumentException if old password is incorrect
     */
    @Transactional
    public void changePassword(ChangePasswordRequest request) {
        User user = getCurrentUser();

        // Verify old password
        if (!passwordEncoder.matches(request.oldPassword(), user.getPassword())) {
            throw new IllegalArgumentException("Current password is incorrect");
        }

        // Validate new password is different
        if (request.oldPassword().equals(request.newPassword())) {
            throw new IllegalArgumentException("New password must be different from current password");
        }

        // Update password
        user.setPassword(passwordEncoder.encode(request.newPassword()));
        userRepository.save(user);
    }
}

File: enterprise-starter-kit-frontend/src/app/services/profile.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';

import { environment } from '../../environments/environment';
import { TokenStorageService } from './token-storage.service';

/**
 * User profile data matching the backend ProfileResponse
 */
export interface UserProfile {
  id: number;
  username: string;
  email: string;
  fullName: string | null;
  phoneNumber: string | null;
  avatarUrl: string | null;
  preferredLanguage: string;
}

/**
 * Request for updating profile
 */
export interface ProfileUpdateRequest {
  fullName?: string;
  phoneNumber?: string;
  preferredLanguage?: string;
}

/**
 * Request for changing password
 */
export interface ChangePasswordRequest {
  oldPassword: string;
  newPassword: string;
}

/**
 * Service for managing user profile operations
 */
@Injectable({
  providedIn: 'root'
})
export class ProfileService {
  private readonly baseUrl = `${environment.apiUrl}/profile`;
  private readonly LANGUAGE_KEY = 'preferred_language';

  constructor(
    private readonly http: HttpClient,
    private readonly tokenStorage: TokenStorageService
  ) {}

  /**
   * Gets the current user's profile
   */
  getProfile(): Observable<UserProfile> {
    return this.http.get<UserProfile>(this.baseUrl);
  }

  /**
   * Updates the current user's profile
   */
  updateProfile(request: ProfileUpdateRequest): Observable<UserProfile> {
    return this.http.put<UserProfile>(this.baseUrl, request).pipe(
      tap((profile) => {
        // Store preferred language in localStorage
        if (profile.preferredLanguage) {
          this.savePreferredLanguage(profile.preferredLanguage);
        }
      })
    );
  }

  /**
   * Changes the current user's password
   */
  changePassword(request: ChangePasswordRequest): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(`${this.baseUrl}/password`, request);
  }

  /**
   * Saves the preferred language to localStorage
   */
  savePreferredLanguage(language: string): void {
    localStorage.setItem(this.LANGUAGE_KEY, language);
  }

  /**
   * Gets the preferred language from localStorage
   */
  getPreferredLanguage(): string {
    return localStorage.getItem(this.LANGUAGE_KEY) || 'en';
  }

  /**
   * Removes the preferred language from localStorage
   */
  clearPreferredLanguage(): void {
    localStorage.removeItem(this.LANGUAGE_KEY);
  }
}
